Attempt 2: Message Boogaloo

Nyefari opened the pull request:
Second Attempt at adding messages DB. Do not approve PR. Still needs try/catch in the cron job and various error handling. Submitting PR only so you can review edits to ensure other changes you wanted were made.

Added FinalUpdateMessages() method to messages which is run by the endcurrentchallenge and clears old challenge's messages from the table. Corrected various issues & typos noted by Martin in previous PR attempt to main repo.

Nyefari committed:
Merge pull request #1 from martindevans/master

Merging in Martin's Changes

Nyefari committed:
Merge pull request #2 from martindevans/master

Merging Changes by Martin on main May 29th

Nyefari committed:
Attempt to create DB for messages and watch it for messages to update. I don't know how to deal with type casts and possible nullability :/

Nyefari committed:
Should have been in last commit

Nyefari committed:
Second Attempt at adding messages DB. Do not approve PR. Still needs try/catch in the cron job and various error handling.

Added FinalUpdateMessages() method to messages which is run by the endcurrentchallenge and clears old challenge's messages from the table. Corrected various issues & typos noted by Martin in previous PR attempt to main repo.

martindevans commented:
I replied to some of your comments on the previous PR with some tips. I'll do another full review on this one too :)

martindevans commented on YololCompetition/Program.cs:
```diff
@@ -39,6 +39,9 @@ public static async Task Main(string[] args)
             await commands.AddModulesAsync(Assembly.GetExecutingAssembly(), provider);
 
+            var messages = provider.GetService<IMessages>();
+            messages.StartMessageWatch();
```
You should do this after starting the bot, just to ensure the message service isn't trying to interact with message before the bot is connected to Discord.

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,148 @@
+        public async Task FinalUpdateMessages()
+        {
+            //Get Current Challenge
+            var c = _challenges.GetCurrentChallenge().Result;
```
use `await _challenges.GetCurrentChallenge()` instead of `Result`. You'll need to make this an `async` method for that.

Nyefari replied:
I was meaning to go back to this and see why it wanted me to use result. Resolved thanks

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,148 @@
+            var c = _challenges.GetCurrentChallenge().Result;
+            if (c == null)
+            {                
+                //Insert Error Handling thing here.
```
Just log a message with `Console.WriteLine` and return is good enough :)

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,148 @@
+                        }                        
+                        break;
+                    default:
+                        //Insert Error Handling thing here
```
Again, just `Console.WriteLine` it.

martindevans commented on YololCompetition/Services/Messages/DBMessages.cs:
```diff
@@ -0,0 +1,148 @@
+            return new SqlAsyncResult<Message>(_database, PrepareQuery, ParseMessage);
+        }
+
+        public async Task FinalUpdateMessages()
```
Most of the code between `FinalUpdateMessages` and `StartMessageWatch` is duplicated, can it be put into another method which they both call into?

martindevans replied:
There's a bit of a race between `FinalUpdateMessages` and `StartMessageWatch` - if `StartMessageWatch` runs when `FinalUpdateMessages` is currently running, what will happen? I don't _think_ it's a problem, but maybe give it some thought or prevent the race.

martindevans commented on YololCompetition/Services/Messages/IMessages.cs:
```diff
@@ -0,0 +1,34 @@
+
+    public enum MessageType : uint {
+        Current, 
+        Leaderboard}
```
`}` should be on the next line.

Nyefari committed:
Implemented Try/Catch Blocks and Console.WriteLine() as error messages.

Nyefari committed:
Fixed Typo in GetCurrentMessages().  Added newline to bottom of IMessages because git had a little error sign saying there wasn't one.

Nyefari reviewed (commented):
Looks good to me. I'm not seeing any other typos in my SQL.

Nyefari committed:
Got rid of FinalUpdateMessage().
Created UpdateCurrentMessage(message, challenge) and updated RemoveMessage and GetMessages functions to be more useful (imo).

Nyefari commented on YololCompetition/Program.cs:
```diff
@@ -39,11 +39,15 @@ public static async Task Main(string[] args)
             Console.WriteLine("Bot Started");
             await provider.GetRequiredService<IScheduler>().Start();
+
+            messages.StartMessageWatch();
```
Moved the startmessagewatch to below where the bot has announced that it is started.

martindevans commented on YololCompetition/Services/Challenge/DbChallenges.cs:
```diff
@@ -24,11 +25,13 @@ private enum Status
+        private readonly IMessages _messages;
 
-        public DbChallenges(IDatabase database, Configuration config)
+        public DbChallenges(IDatabase database, Configuration config, IMessages messages)
```
This takes `IMessages` as a parameter, but doesn't seem to use it any more.

Nyefari replied:
true true

Nyefari committed:
fixing Imessage being referenced in DBChallenges since that isn't needed anymore.

Nyefari committed:
fixed times I used single quote rather than backtick

martindevans commented:
Merged manually
